docs/concurrency-adoption-guidelines.md

```diff
@@ -0,0 +1,179 @@
+# Swift Concurrency adoption guidelines for Swift Server Libraries
+
+This writeup attempts to provide a set of guidelines to follow by authors of server-side swift libraries. Specifically a lot of the discussion here revolves around what to do about existing APIs and libraries making extensive use of Swift NIO’s EventLoopFuture and related types.
+
+Swift Concurrency is a multi-year effort. It is very valuable for the server community to participate in this multi-year adoption of the concurrency features, one by one, and provide feedback while doing so. As such, we should not hold off adopting concurrency features until Swift 6 as we may miss valuable opportunity to improve the concurrency model.
+
+In 2021 we saw structured concurrency and actors. Now is a great time to provide APIs using those primitives. In the future we will see fully checked Swift concurrency. This will come with breaking changes. For this reason adopting the new concurrency features can be split into two phases.
+
+
+## What you can do right now
+
+### API Design
+
+Firstly, existing libraries should strive to add `async` functions where possible to their user-facing “surface” APIs in addition to existing `*Future` based APIs wherever possible. These additive APIs can be gated on the Swift version and can be added without breaking existing users, for example like this:
```

Comment:
do we expect anyone who this guide applies to is using any future type besides `EventLoopFuture`?

Reply:
```suggestion
Firstly, existing libraries should strive to add `async` functions where possible to their user-facing “surface” APIs in addition to existing `*Future` based APIs wherever possible. These additive APIs can be gated on the Swift version and can be added without breaking existing users' code, for example like this:
```

Reply:
Hmm... I guess there MIGHT BE some library using the other future libraries but honestly I've not seen one for server focus that would use the other ones (PromiseKit etc), they're usually a bit iOS "tainted" hm...